compiler/semexprs.nim

```diff
@@ -1060,7 +1060,7 @@ proc builtinFieldAccess(c: PContext, n: PNode, flags: TExprFlags): PNode =
     of tyTypeParamsHolders:
       return readTypeParameter(c, ty, i, n.info)
     of tyObject, tyTuple:
-      if ty.n != nil and ty.n.kind == nkRecList:
+      if (not isTypeExpr(n.sons[0])) and ty.n != nil and ty.n.kind == nkRecList:
```

Comment:
All I really know is that it fixes the issue and it doesn't seem to break
any tests. I don't understand the compiler well enough to reason about the
code globally.

But as far as I can tell the code is not necessary to implement the
semantics of type(x.name) (as seems to be the purpose), but messes up the
interpretation of type(x).name if x is an object containing the field name.

But it could be totally wrong.

It would be nice if there were more comments in the compiler. It's a bit
hard for an outsider to get into.

On Thursday, 14 January 2016, Andreas Rumpf <notifications@github.com>
wrote:

> In compiler/semexprs.nim
> <https://github.com/nim-lang/Nim/pull/3712#discussion_r49706963>:
>
> > @@ -1060,7 +1060,7 @@ proc builtinFieldAccess(c: PContext, n: PNode, flags: TExprFlags): PNode =
> >      of tyTypeParamsHolders:
> >        return readTypeParameter(c, ty, i, n.info)
> >      of tyObject, tyTuple:
> > -      if ty.n != nil and ty.n.kind == nkRecList:
> > +      if (not isTypeExpr(n.sons[0])) and ty.n != nil and ty.n.kind == nkRecList:
>
> Why is that necessary?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/nim-lang/Nim/pull/3712/files#r49706963>.
>